### ISSUE TITLE
Ask Permission app flow question

### ISSUE THREAD

**Issue by Jarvey** [phonegap/phonegap-plugin-push#1802]
### Expected Behaviour

When a user starts the app for the first time it is immediately confronted wit the alerts permission popup. To postpone this popup until the user has completed the app setup I have put the PushNotification.init() in a angular service so I can call it after the setup is completed.

What I'm  currently doing:
I have moved the PushNotification.init() in to a angular service.
If the the app has a cold start and the app setup is completed I run my notification init service setup.pushInit(); to initialise the push plugin (what normally happens).

When the app is started for the first time (setup is not completed yet) I wait with triggering the setup.pushInit(); until the user has finished the setup.

I expect that the permission popup is presented after the setup and that the rest is working as normal. 

### Actual Behaviour

The permission popup is presented after the setup so that is as expected.
Notifications are still received, the problem is that  `push.on('notification', function(data) {}` is not triggered anymore when notification is received in background, foreground etc. because the notification data is always undefined. 

### Reproduce Scenario (including but not limited to)

Android & iOS

#### Platform and Version (eg. Android 5.0 or iOS 9.2.1)

iOS 10
Android 5.0.1

#### (Android) What device vendor (e.g. Samsung, HTC, Sony...)

Samsung S4

#### Cordova CLI version and cordova platform version

    cordova --version                                    # 7.0.1
    cordova platform version android          # 6.2.2
    cordova platform version iOS                 # 4.4.0

#### Plugin version

    1.10.2

#### Sample Code that illustrates the problem

```
.run(function($cordovaNativeStorage, notification) {
  $ionicPlatform.ready(function() {
    $cordovaNativeStorage.getItem('setupCompleted').then(function (setup) {
      if (setup) {
        setup.pushInit();
      };
    }, callbackError);
  });
});	

//Push notification service
.service('setup', function() {
  //init function
  this.pushInit = function() {   
    //initialize push service 
    var push = PushNotification.init({
      ios: {
        alert: "true",
	badge: "true",
	sound: "true",
	clearBadge: "false"
      },
      android: {
        senderID: "12345",
	sound: "true",
	vibrate: "false",
	clearBadge: "false",
	iconColor: "#34363a"
      }
    });		
		
    push.on('registration', function(data) {
       //store devicetoken
    });

    push.on('notification', function(data) {
       //console.log(data)
    });

    push.on('error', function(e) {
      console.log(e.message);
    });
  }
});
```

**Comment by macdonst** [comment]
@Jarvey I confess that I do not know. It seems like you are doing things right but I'm not very well versed in Angular.

**Comment by fredgalvao** [comment]
Is this the only place you try to call `pushInit()`?:
```js
  $ionicPlatform.ready(function() {
    $cordovaNativeStorage.getItem('setupCompleted').then(function (setup) {
      if (setup) {
        setup.pushInit();
      };
    }, callbackError);
  });
```

If so, you might be creating a leak where, on your app's initialization, if the setup is not completed by the time that code runs, you'll not initialize the push plugin, and thus making you not receive any event until the next initialization. If this is the case, you should call `setup.pushInit()` whenever that `setupCompleted` entry gets updated.


Now, when you say:
> Notifications are still received, the problem is that push.on('notification', function(data) {} is not triggered anymore when notification is received in background, foreground etc. because the notification data is always undefined.

Can you describe exactly what happens and what is the value of `on('notification', function(d){console.log(d)}`?:
- when a notification arrives and the app is in the foreground
- when a notification arrives and the app is in the background


Also, you missed the `Sample Push Data Payload` and `Logs taken while reproducing problem` sections of the issue template, which are crucial to figuring out your issue.

**Comment by macdonst** [comment]
Closing, inactivity. Please comment/re-open if you have more details.

**Comment by dfliess** [comment]
Hi, I'm having a similar issue. If i do not run init() on app start I do not receive notifications.

@Jarvey have you found a solution ?

**Comment by dfliess** [comment]
If I put the init on my startup on iOS the user gets a prompt that's very invasive. I want to prompt only after the user makes an action (for example login), but the next time the user enters the app I don't want him to login again, but I want to receive the notifications events.
In this case I stored a variable in state and checked that on app startup, if it was not initialized I do not init the plugin, and changed that state after the action completes, so next startup it listens for notifications.
Maybe it would be useful to ask for hasPermission without initializing.

**Comment by fredgalvao** [comment]
Regardless of when you call `.init()`, which will depend on your use case, some things need to be clear:
- your app will not be able to react to any notification delivered **until you call `.init()`**
- everytime your app starts (when it was not in the background before) you'll need to call `.init()` again
